### PR TITLE
install: add --delete-karg and --append-karg

### PR DESCRIPTION
**Pull request by jlebon**
This allows users to delete and append kernel arguments from the default
set. The major use cases for this are (1) avoiding an additional reboot
which might be costly in some situations, and (2) making it easier to
troubleshoot the system on firstboot.

For more details, see
https://github.com/coreos/fedora-coreos-tracker/issues/533#issuecomment-642234015.

Closes: https://github.com/coreos/fedora-coreos-tracker/issues/533

### CODE REVIEW

File: src/install.rs [hunk 1]
```diff
@@ -254,6 +259,31 @@ fn write_firstboot_kargs(mountpoint: &Path, args: &str) -> Result<()> {
     Ok(())
 }
 
+/// Write persistent kernel arguments.
```
Comment (cgwalters) [inline src/install.rs:262]: One thing we need to keep in mind is this needs to trigger a rerun of `zipl` on s390x.

Hmm...I am increasingly thinking what we want is like a special systemd unit for zipl that checks the mtime of the BLS config and runs `zipl` if it's changed or something like that.

**Comment by brokenjacobs** [comment]
Is there also a way to remove persistent arguments as well? (console=ttyS0,1152008n1).

It's a good idea in my opinion to emulate the various arguments to rpm-ostree kargs. They are all there for a reason.

**Comment by lucab** [comment]
I like this but also have a bad feeling about it; I fear this is going to trip many people.

In particular, it looks like the logic implemented here is really an `--append-persistent-kargs`. Also, it implicitly assumes that users already know what are the current kargs (is that a static set?) and that previous values can be overridden in a last-key-wins ways (I know that's not the case generally, see for example hugepages).

File: src/install.rs (hunk 1 above)
Comment (nikita-dubrovskii) [reply src/install.rs:262]: Indeed, after installation we do write some s390x specific args parsed from `/proc/cmdline`

File: src/install.rs [hunk 2]
```diff
@@ -254,6 +259,31 @@ fn write_firstboot_kargs(mountpoint: &Path, args: &str) -> Result<()> {
...
+fn write_persistent_kargs(mountpoint: &Path, args: &str) -> Result<()> {
+    eprintln!("Writing persistent kernel arguments");
+
+    edit_bls_entries(mountpoint, |orig_contents: String| {
```
Comment (lucab) [inline src/install.rs:266]: For this and the other similar closure below, what about moving it out of this scope instead? I'd be really happier to see both covered with a few positive and negative unit tests.

**Comment by jlebon** [comment]
> In particular, it looks like the logic implemented here is really an `--append-persistent-kargs`.

Yeah, see https://github.com/coreos/fedora-coreos-tracker/issues/533#issuecomment-644319795 related to this. Will rename the switch.

> Also, it implicitly assumes that users already know what are the current kargs (is that a static set?)

Hmm, though that's no different from eventual kargs-via-Ignition support, right?

>  and that previous values can be overridden in a last-key-wins ways (I know that's not the case generally, see for example hugepages).

Yes, agreed (turns out `console` is another such example as mentioned in the linked comment above). Combined with `--delete-persistent-kargs`, I think it provides enough flexibility.

(See also https://github.com/ostreedev/ostree/issues/1859 -- we'll need to make sure that the implementation of kargs-via-Ignition is ordering aware).

**Comment by jlebon** [comment]
OK, split out prep patches in https://github.com/coreos/coreos-installer/pull/271!

File: src/install.rs (hunk 2 above)
Comment (jlebon) [reply src/install.rs]: This is part of https://github.com/coreos/coreos-installer/pull/271.

File: src/install.rs (hunk 1 above)
Comment (jlebon) [reply src/install.rs]: Ahh yup, goot catch. Hmm, let's circle back on this once https://github.com/coreos/coreos-installer/pull/153 merges? It'll make it easier to run `zipl`. I can open an issue once this merges.

Comment (nikita-dubrovskii) [reply src/install.rs]: Sorry, haven't seen this message begore. So, i've already started working on patch to remove all logic in #153 `zipl.rs` in favour of using `--append-karg`.

**Pull request merged**